src/ray/raylet/worker_pool.cc

```diff
@@ -715,11 +716,13 @@ void WorkerPool::TryKillingIdleWorkers() {
   running_size -= pending_exit_idle_workers_.size();
   // Kill idle workers in FIFO order.
   for (const auto &idle_pair : idle_of_all_languages_) {
-    if (running_size <= static_cast<size_t>(num_workers_soft_limit_)) {
+    const auto &idle_worker = idle_pair.first;
+    const auto &job_id = idle_worker->GetAssignedJobId();
+    if (running_size <= static_cast<size_t>(num_workers_soft_limit_) &&
+        !finished_jobs_.count(job_id)) {
```

Comment:
Does this mean that even the job has a lot of idle workers, if the job hasn't finished, we'll never kill any idle work?

Reply:
I think this is right? It should be:
few workers && job not finished -> don't kill
lots of workers -> kill
job finished -> kill

Reply:
Sorry, my bad. You're right.

Reply:
I still feel some confusion here. Let me know if I'm wrong. If we got this code path, it means the workers will drain all the references before exiting right. Is it necessary to check whether the job finished or not?

Think about this case, there is another job running with a lot of workers, for the current job, although it hasn't finished, we still kill the worker which is the same as before?

Reply:
Hmm yeah I added this to take care of the case where a job has finished but its workers are still around. Assuming no other workers will get started, the finished job's workers won't get killed because they're still under the soft limit.

It's not really necessary if the cluster is actively used, though, since another job will start more workers and then we should enter the idle-killing cleanup.

Reply:
Maybe I should just add a comment here to explain why we have the finished check?

Reply:
Ah, got it. It makes sense! I think I went to the wrong direction ~~(we don't want to kill idle workers when the job is active)~~

Reply:
Okay! Yeah, this bit of logic is confusing. I added a comment, BTW!